### PR TITLE
[JENKINS-30626] Handle pings from organization webhooks

### PR DESCRIPTION
**Pull request by jglick**
[JENKINS-30626](https://issues.jenkins-ci.org/browse/JENKINS-30626)

[Reference](https://developer.github.com/v3/orgs/hooks/)

@reviewbybees esp. @recena who discovered this bug.

### CODE REVIEW

**Comment by recena** [comment]
Of course! :bee: and :+1:

**Comment by oleg-nenashev** [comment]
:bee: , but seems to be significant enough to have a JIRA issue. Does the original code fail with NPE?

**Comment by amuniz** [comment]
:bee:

**Comment by lanwen** [comment]
Please add a test with such payload same as for the ping.

File: src/main/java/org/jenkinsci/plugins/github/webhook/subscriber/PingGHEventSubscriber.java [hunk 1]
```diff
@@ -10,7 +10,7 @@
 import java.util.Set;
 
 import static com.google.common.collect.Sets.immutableEnumSet;
-import static net.sf.json.JSONObject.fromObject;
+import net.sf.json.JSONObject;
```
Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github/webhook/subscriber/PingGHEventSubscriber.java]: then it should be not in static block of imports

File: src/main/java/org/jenkinsci/plugins/github/webhook/subscriber/PingGHEventSubscriber.java [hunk 2]
```diff
@@ -52,8 +52,18 @@ protected boolean isApplicable(Job<?, ?> project) {
      */
     @Override
     protected void onEvent(GHEvent event, String payload) {
-        // something like <https://github.com/bar/foo>
-        String repo = fromObject(payload).getJSONObject("repository").getString("url");
-        LOGGER.info("{} webhook received from repo <{}>!", event, repo);
+        JSONObject parsedPayload = JSONObject.fromObject(payload);
```
Comment (lanwen) [inline src/main/java/org/jenkinsci/plugins/github/webhook/subscriber/PingGHEventSubscriber.java]: why you remove static import?

Comment (KostyaSha) [inline src/main/java/org/jenkinsci/plugins/github/webhook/subscriber/PingGHEventSubscriber.java]: for using JSONObject

**Comment by lanwen** [comment]
:+1: thanks. I'll going to squash this and merge manually

**Comment by recena** [comment]
@lanwen Why closed? /cc @jglick

**Comment by KostyaSha** [comment]
Please read what GH wrote: `@jglickjglick closed this in 03b1c87 8 minutes ago`

**Comment by recena** [comment]
@KostyaSha Thanks. I have to learn why this way is better (than traditional merge).

**Comment by KostyaSha** [comment]
It not better, GH closes relevant PR when changes committed in target branch directly. Faster fix ourself instead of commenting and waiting for changes from PR author.

**Comment by recena** [comment]
@KostyaSha ah ok. Thanks for your time.

File: src/main/java/org/jenkinsci/plugins/github/webhook/subscriber/PingGHEventSubscriber.java [hunk 3]
```diff
@@ -52,8 +53,18 @@ protected boolean isApplicable(Job<?, ?> project) {
      */
     @Override
     protected void onEvent(GHEvent event, String payload) {
-        // something like <https://github.com/bar/foo>
-        String repo = fromObject(payload).getJSONObject("repository").getString("url");
-        LOGGER.info("{} webhook received from repo <{}>!", event, repo);
+        JSONObject parsedPayload = fromObject(payload);
+        JSONObject repository = parsedPayload.optJSONObject("repository");
+        if (repository != null) {
+            // something like <https://github.com/bar/foo>
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github/webhook/subscriber/PingGHEventSubscriber.java]: Not sure it matters, but not that this is not actually the form of the URL; it is something involving `api.github.com`.

**Comment by jglick** [comment]
> GH closes relevant PR when changes committed in target branch

Well, when the special `closes` keyword is used in the squashed commit comment. It only treats the PR as having the merged status if the head commit of the PR actually becomes an ancestor of the head commit of the target branch, which will not happen when you squash.

**Comment by KostyaSha** [comment]
Special keyword is not required. When diff has nothing GH closes it.

File: src/main/java/org/jenkinsci/plugins/github/webhook/subscriber/PingGHEventSubscriber.java (hunk 3 above)
Comment (recena) [inline src/main/java/org/jenkinsci/plugins/github/webhook/subscriber/PingGHEventSubscriber.java]: @jglick For that, there are `GHEventsSubscriber` [validating](https://github.com/jenkinsci/github-plugin/blob/master/src/main/java/org/jenkinsci/plugins/github/webhook/subscriber/DefaultPushGHEventSubscriber.java#L73) the URL.

Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github/webhook/subscriber/PingGHEventSubscriber.java]: My point was just that the comment is wrong.

**Pull request closed without merging**